src/tcache.c

```diff
@@ -169,6 +167,137 @@ tcache_gc_small(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
 	     tcache_slow->lg_fill_div[szind]) > 1) {
 		tcache_slow->lg_fill_div[szind]++;
 	}
+
+	if (nflush == ncached)	goto label_flush;
+
+	/* Query arena binshard to get heuristic locality info. */
+	tsdn_t *tsdn = tsd_tsdn(tsd);
+	bin_t *bin = arena_bin_choose(tsdn, tcache_slow->arena, szind, NULL);
+	assert(bin != NULL);
+
+	void *addr = NULL;
+	malloc_mutex_lock(tsdn, &bin->lock);
+	edata_t *slab = bin->slabcur == NULL ?
+	    edata_heap_first(&bin->slabs_nonfull) : bin->slabcur;
+	assert(slab != NULL || edata_heap_empty(&bin->slabs_nonfull));
+	if (slab != NULL) {
+		addr = edata_addr_get(slab);
+	}
+	assert(addr != NULL || slab == NULL);
+	malloc_mutex_unlock(tsdn, &bin->lock);
+	if (addr == NULL)	goto label_flush;
+
+	/* The slab address range that the queried addr belongs to. */
+	uintptr_t slab_min = (uintptr_t)addr;
+	uintptr_t slab_max = slab_min + bin_infos[szind].slab_size;
+	/* The exp_grow->next starts from HUGEPAGE. */
+	uintptr_t addr_min = slab_min > HUGEPAGE ? (slab_min - HUGEPAGE) : 0;
+	uintptr_t addr_max = slab_min < ((uintptr_t)(-1) - HUGEPAGE) ?
+	    (slab_min + HUGEPAGE) : (uintptr_t)(-1);
+	cache_bin_sz_t nremote = 0, nremote_hard = 0;
+	void **head = cache_bin->stack_head;
+	/* Scan the entire bin to count the number of remote pointers. */
+	for (void **cur = head; cur < head + ncached; cur++) {
+		assert(*cur != NULL);
+		uintptr_t addr_cur = (uintptr_t)*cur;
+		nremote += (cache_bin_sz_t)(addr_cur < addr_min ||
+		    addr_cur >= addr_max);
+		nremote_hard += (cache_bin_sz_t)(addr_cur < slab_min ||
+		    addr_cur >= slab_max);
+	}
+	assert(nremote_hard >= nremote && nremote_hard <= ncached);
+	/*
+	 * The number of ptrs outside of [addr_min, addr_max) is less than
+	 * the intended flush amount, shrink the range to [slab_min, slab_max).
+	 */
+	if (nremote < nflush) {
+		nremote = nremote_hard;
+		addr_min = slab_min;
+		addr_max = slab_max;
+	}
+	/*
+	 * Update the nflush to the larger value between the intended flush count
+	 * and the number of remote ptrs.
+	 */
+	nflush = nremote > nflush ? nremote : nflush;
+	assert(nflush < ncached || nremote == ncached);
```

Comment:
I don't follow this assertion. We have nflush <= ncached always. So this is saying, if (nflush == ncached), everything must be remote?

Reply:
In L171, we have if (nflush == ncached) goto label_flush; so that if the number of items to be flushed is already equal to ncached, we don't need to go through the locality route at all.

So when we enter the logic of locality flush, nflush should be less than ncached. Unless after locality check, that all the items are remote, and in that case we round the nflush up to nremote, leading nflush == ncached.